lib/chef/resource/macos_userdefaults.rb

```diff
@@ -28,99 +30,180 @@ class MacosUserDefaults < Chef::Resource
-                   "NSGlobalDomain"
-                 else
-                   "'#{new_resource.domain}'"
-                 end
-
-          cmd << "'#{new_resource.key}'" if new_resource.key
-          value = new_resource.value
-          type = new_resource.type.empty? ? value_type(value) : new_resource.type
-          # creates a string of Key1 Value1 Key2 Value2...
-          value = value.map { |k, v| "\"#{k}\" \"#{v}\"" }.join(" ") if type == "dict"
-          if type == "array"
-            value = value.join("' '")
-            value = "'#{value}'"
+        description "Write the value to the specified domain/key."
+
+        converge_if_changed do
+          cmd = defaults_modify_cmd("write")
+          Chef::Log.debug("Updating defaults value by shelling out: #{cmd.join(" ")}")
+
+          if new_resource.user.nil?
+            shell_out!(cmd)
+          else
+            shell_out!(cmd, user: new_resource.user)
           end
-          cmd << "-#{type}" if type
-          cmd << value
+        end
+      end
+
+      action :delete do
+        description "Delete a key from a domain."
+
+        return unless current_resource.key # if it's not there there's nothing to remove
 
-          # FIXME: this should use cmd directly as an array argument, but then the quoting
-          # of individual args above needs to be removed as well.
-          execute cmd.join(" ") do
-            user new_resource.user unless new_resource.user.nil?
+        converge_by("delete domain:#{new_resource.domain} key:#{new_resource.key}") do
+
+          cmd = defaults_modify_cmd("delete")
+          Chef::Log.debug("Removing defaults key by shelling out: #{cmd.join(" ")}")
+
+          if new_resource.user.nil?
+            shell_out!(cmd)
+          else
+            shell_out!(cmd, user: new_resource.user)
           end
         end
       end
 
       action_class do
+        def defaults_modify_cmd(defaults_action)
+          cmd = ["defaults"]
+
+          if new_resource.host == "current"
+            state_cmd.concat(["-currentHost"])
+          elsif new_resource.host # they specified a non-nil value, which is a hostname
+            state_cmd.concat(["-host", new_resource.host])
+          end
+
+          cmd.concat([defaults_action, new_resource.domain, new_resource.key])
+          cmd.concat(processed_value) if defaults_action == "write"
+          cmd.prepend("sudo") if new_resource.sudo
+          cmd
+        end
+
+        #
+        # convert the provided value into the format defaults expects
+        #
+        # @return [array] array of values starting with the type if applicable
+        #
+        def processed_value
+          type = new_resource.type || value_type(value)
+
+          # when dict this creates an array of values ["Key1", "Value1", "Key2", "Value2" ...]
+          cmd_vals = [ type == "dict" ? new_resource.value.flatten : new_resource.value ]
+          cmd_vals.prepend("-#{type}") if type
```

Comment:
Spellcheck does not like this variable name:

```
/workdir/lib/chef/resource/macos_userdefaults.rb:196:15 - Unknown word (vals)
/workdir/lib/chef/resource/macos_userdefaults.rb:197:15 - Unknown word (vals)
```

Reply:
Also if `type` is nil here this method will return `nil`.

Reply:
I'm noodling on how to destroy this whole set of methods. It's all pretty fragile.

Reply:
bit weird that you slurp up all the values and then prepend the type rather than just appending.

prepend is a nifty trick but it should probably be very lightly used

Reply:
It avoided having to create an empty array, then optionally concat the type in, then toss the values in.